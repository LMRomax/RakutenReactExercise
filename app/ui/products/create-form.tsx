'use client'

import { ProductField } from '@/app/lib/definitions';
import { createProduct } from '@/app/lib/action';
import { useFormState } from 'react-dom';

export default function Form() {
    const initialState = { message: '', errors: {} }
    const [state, dispatch] = useFormState(createProduct, initialState);

    return (
        <div className="w-full flex items-start justify-start">
            <div className="bg-[#f4f4f4] w-full p-6 rounded-lg md:w-10/12 xl:w-6/12">
                <form action={dispatch}>
                    <div className="mb-4">
                        <label htmlFor="brand" className="mb-2 block text-sm font-medium">
                            Brand
                        </label>
                        <div className="relative mt-2 rounded-md">
                            <div className="relative">
                                <input
                                    id="brand"
                                    name="brand"
                                    type="text"
                                    placeholder="Enter a brand"
                                    className="peer block w-full rounded-md border border-gray-200 px-3 py-2 text-sm outline-2 placeholder:text-gray-500 focus:ring-[#bf0000] focus:border-transparent"
                                />
                            </div>
                        </div>

                        <div id="brand-error" aria-live="polite" aria-atomic="true">
                            {state.errors?.brand &&
                                state.errors.brand.map((error: string) => (
                                    <p className="mt-2 text-sm text-red-500" key={error}>
                                        {error}
                                    </p>
                                ))}
                        </div>
                    </div>

                    <div className="mb-4">
                        <label htmlFor="title" className="mb-2 block text-sm font-medium">
                            Title
                        </label>
                        <div className="relative mt-2 rounded-md">
                            <div className="relative">
                                <input
                                    id="title"
                                    name="title"
                                    type="text"
                                    placeholder="Enter a title"
                                    className="peer block w-full rounded-md border border-gray-200 px-3 py-2 text-sm outline-2 placeholder:text-gray-500 focus:ring-[#bf0000] focus:border-transparent"
                                />
                            </div>
                        </div>

                        <div id="title-error" aria-live="polite" aria-atomic="true">
                            {state.errors?.title &&
                                state.errors.title.map((error: string) => (
                                    <p className="mt-2 text-sm text-red-500" key={error}>
                                        {error}
                                    </p>
                                ))}
                        </div>
                    </div>

                    <div className="mb-4">
                        <label htmlFor="href" className="mb-2 block text-sm font-medium">
                            Url of product
                        </label>
                        <div className="relative mt-2 rounded-md">
                            <div className="relative">
                                <input
                                    id="href"
                                    name="href"
                                    type="text"
                                    placeholder="Enter an url for the product"
                                    className="peer block w-full rounded-md border border-gray-200 px-3 py-2 text-sm outline-2 placeholder:text-gray-500 focus:ring-[#bf0000] focus:border-transparent"
                                />
                            </div>
                        </div>

                        <div id="href-error" aria-live="polite" aria-atomic="true">
                            {state.errors?.href &&
                                state.errors.href.map((error: string) => (
                                    <p className="mt-2 text-sm text-red-500" key={error}>
                                        {error}
                                    </p>
                                ))}
                        </div>
                    </div>

                    <div className="mb-4">
                        <label htmlFor="image" className="mb-2 block text-sm font-medium">
                            Image of product
                        </label>
                        <div className="relative mt-2 rounded-md">
                            <div className="relative">
                                <input
                                    id="image"
                                    name="image"
                                    type="text"
                                    placeholder="Entre the URL of the product"
                                    className="peer block w-full rounded-md border border-gray-200 px-3 py-2 text-sm outline-2 placeholder:text-gray-500 focus:ring-[#bf0000] focus:border-transparent"
                                />
                            </div>
                        </div>

                        <div id="image-error" aria-live="polite" aria-atomic="true">
                            {state.errors?.image &&
                                state.errors.image.map((error: string) => (
                                    <p className="mt-2 text-sm text-red-500" key={error}>
                                        {error}
                                    </p>
                                ))}
                        </div>
                    </div>

                    <div className="mb-4">
                        <label htmlFor="newprice" className="mb-2 block text-sm font-medium">
                            Price of new product
                        </label>
                        <div className="relative mt-2 rounded-md">
                            <div className="relative">
                                <input
                                    id="newprice"
                                    name="newprice"
                                    type="text"
                                    placeholder="Entre the URL of the product"
                                    className="peer block w-full rounded-md border border-gray-200 px-3 py-2 text-sm outline-2 placeholder:text-gray-500 focus:ring-[#bf0000] focus:border-transparent"
                                />
                            </div>
                        </div>

                        <div id="newprice-error" aria-live="polite" aria-atomic="true">
                            {state.errors?.newprice &&
                                state.errors.newprice.map((error: string) => (
                                    <p className="mt-2 text-sm text-red-500" key={error}>
                                        {error}
                                    </p>
                                ))}
                        </div>
                    </div>

                    <div className="mb-4">
                        <label htmlFor="usedprice" className="mb-2 block text-sm font-medium">
                            Price of used product
                        </label>
                        <div className="relative mt-2 rounded-md">
                            <div className="relative">
                                <input
                                    id="usedprice"
                                    name="usedprice"
                                    type="text"
                                    placeholder="Entre the URL of the product"
                                    className="peer block w-full rounded-md border border-gray-200 px-3 py-2 text-sm outline-2 placeholder:text-gray-500 focus:ring-[#bf0000] focus:border-transparent"
                                />
                            </div>
                        </div>

                        <div id="usedprice-error" aria-live="polite" aria-atomic="true">
                            {state.errors?.usedprice &&
                                state.errors.usedprice.map((error: string) => (
                                    <p className="mt-2 text-sm text-red-500" key={error}>
                                        {error}
                                    </p>
                                ))}
                        </div>
                    </div>

                    <div className="flex items-center justify-center mt-12">
                        <button className="bg-[#bf0000] p-3 text-white rounded-lg" type="submit">
                            Create product
                        </button>
                    </div>
                </form>
            </div>
        </div>
    )
}
