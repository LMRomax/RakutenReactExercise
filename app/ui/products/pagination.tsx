'use client';

import { ArrowLeftIcon, ArrowRightIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';
import Link from 'next/link';
import { generatePagination } from '@/app/lib/utils';
import { usePathname, useSearchParams } from 'next/navigation';
import { current } from '@reduxjs/toolkit';

export default function Pagination({ totalPages }: { totalPages: number }) {
  const pathname = usePathname();
  const searchParams = useSearchParams();
  const currentPage = Number(searchParams.get('page')) || 1;

  const allPages = generatePagination(currentPage, totalPages);

  const createPageURL = (pageNumber: number | string) => {
    const params = new URLSearchParams(searchParams);
    params.set('page', pageNumber.toString());
    return `${pathname}?${params.toString()}`;
  }

  return (
    <>
      <div className="w-full flex items-center justify-center inline-flex mt-8">
        <PaginateArrow
          direction='left'
          href={createPageURL(currentPage - 1)}
          isDisabled={currentPage <= 1}
        />

        <div className="flex -space-x-px">
          {allPages.map((page, index) => {
            let position: 'first' | 'last' | 'single' | 'middle' | undefined;

            if(index === 0) position = 'first';
            if(index === allPages.length - 1) position = 'last';
            if(allPages.length === 1) position = 'single';
            if(page === '...') position = 'middle';

            return (
              <PaginationNumber 
                key={page}
                page={page}
                position={position}
                href={createPageURL(page)}
                isActive={currentPage === page}
              />
            );
          })}
        </div>

        <PaginateArrow
          direction='right'
          href={createPageURL(currentPage + 1)}
          isDisabled={currentPage >= totalPages}
        />
      </div>
    </>
  );
}

function PaginationNumber({
  page,
  position,
  href,
  isActive
}: {
  page: number | string,
  position?: 'first' | 'last' | 'single' | 'middle',
  href: string,
  isActive?: boolean
}) {
  const className = clsx(
    'flex h-10 w-10 items-center justify-center text-sm border',
    {
      'z-10 bg-[#bf0000] border-[#bf0000] text-white': isActive,
      'text-[#bf0000] hover:bg-gray-100': !isActive && position !== 'middle',
      'text-gray-300': position === 'middle',
    },
  );

  return isActive && position === 'middle' ? (
    <div className={className}>
      {page}
    </div>
  ) : (
    <Link href={href} className={className}>
      {page}
    </Link>
  );
}

function PaginateArrow({
  href,
  direction,
  isDisabled
}: {
  href: string,
  direction: 'left' | 'right',
  isDisabled?: boolean
}) {
  const className = clsx(
    'flex h-10 items-center justify-center border p-3',
    {
      'pointer-events-none text-gray-300': isDisabled,
      'hover:bg-gray-100': !isDisabled,
      'text-gray-600': direction === 'left' && !isDisabled,
      'border-r-0 rounded-l-md text-gray-400': direction === 'left',
      'text-[#bf0000]/50': direction === 'right' && isDisabled,
      'border-l-0 rounded-r-md text-[#bf0000]': direction === 'right'
    },
  );

  const text = direction === 'left' ? (
    `< Précédent`
  ) : (
    `Suivant >`
  );

  return isDisabled ? (
    <div className={className}>{text}</div>
  ) : (
    <Link href={href} className={className}>
      {text}
    </Link>
  );
}