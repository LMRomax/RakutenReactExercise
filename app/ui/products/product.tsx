import Image from "next/image"
import { fetchFilteredProducts } from '@/app/lib/data';
import { formatCurrency } from "@/app/lib/utils";
import Link from "next/link";

export async function AllProducts({
    currentPage,
    itemPerPage,
    search
}: {
    currentPage: number,
    itemPerPage: number,
    search: string
}) {
    const jsonProducts = await fetchFilteredProducts(currentPage, itemPerPage, search);
    const products = JSON.parse(jsonProducts.toString()) as Array<any>;

    return (
        <div className="w-full grid gap-2 grid-cols-1 md:grid-cols-2 xl:grid-cols-3">
            {products.map((product) => {
                return (
                    <Product
                        key={product.id}
                        image={product.image as string}
                        href={product.href as string}
                        brand={product.brand as string}
                        title={product.title as string}
                        newPrice={product.newprice as number}
                        usedPrice={product.usedprice as number}
                    />
                );
            })}
        </div>
    );
}

export function Product({
    href,
    brand,
    title,
    newPrice,
    usedPrice,
    image
}: {
    href: string,
    brand: string,
    title: string,
    newPrice: number,
    usedPrice: number,
    image: string
}) {
    const isActive = href === null 
    || image === null 
    || brand === null 
    || title === null
    || newPrice === null 
    || usedPrice === null ? false : true;

    return isActive === true ? (
        <Link data-testid="product"
        href={href}
        className="rounded-lg p-[16px] flex flex-row gap-[16px] items-center justify-center border border-slate-300 w-full">
            <div className="flex items-center justify-center w-[120px] h-[120px]">
                <Image
                    src={image}
                    width={0}
                    height={0}
                    sizes="100vw"
                    style={{ width: '100%', height: 'auto' }}
                    alt="Image od product name"
                />
            </div>

            <div>
                <div className="text-[#696969] text-[12px] font-semibold">
                    {brand}
                </div>

                <div className="text-[#333] text-[14px] mb-[8px] font-semibold">
                    {title}
                </div>

                <div className="text-[#BF0000] text-[14px] font-bold">
                    <span className="text-[20px]">{formatCurrency(newPrice)}</span> Neuf
                </div>

                <div className="text-[#333] text-[14px]">
                    Occasion d√©s <span className="text-[20px] font-bold">{formatCurrency(usedPrice)}</span>
                </div>
            </div>
        </Link>
    ) : <></>;
}
