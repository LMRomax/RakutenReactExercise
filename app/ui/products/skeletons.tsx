// Loading animation
const shimmer =
    'before:absolute before:inset-0 before:-translate-x-full before:animate-[shimmer_2s_infinite] before:bg-gradient-to-r before:from-transparent before:via-white/60 before:to-transparent';

export function ProductSkeleton({
    itemPerPage
}: {
    itemPerPage: number
}) {
    return (
        Array(itemPerPage).fill(0).map((el, index) => (
            <div className={`${shimmer} relative overflow-hidden rounded-lg p-[16px] flex flex-row gap-[16px] items-center justify-center border border-slate-300 w-full`}>
                <div className="w-[120px] h-[120px] rounded-lg bg-gray-200"></div>

                <div>
                    <div className="w-[64px] h-[18px] text-[#696969] text-[12px] font-semibold bg-gray-200 mb-2"></div>

                    <div className="w-[64px] h-[42px] text-[#333] text-[14px] mb-[8px] font-semibold bg-gray-200 mb-[8px]"></div>

                    <div className="w-[64px] h-[24px] text-[#BF0000] text-[14px] font-bold bg-gray-200 mb-2"></div>

                    <div className="w-[64px] h-[24px] text-[#333] text-[14px] bg-gray-200 mb-2"></div>
                </div>
            </div>
        ))
    );
}

export function ProductSkeletonWrapper({
    itemPerPage
}: {
    itemPerPage: number
}) {
    return (
        <>
            <div className="w-full grid gap-2 grid-cols-1 md:grid-cols-2 xl:grid-cols-3">
                <ProductSkeleton itemPerPage={itemPerPage} />
            </div>
        </>
    );
}