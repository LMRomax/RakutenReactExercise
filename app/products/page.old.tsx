import { AllProducts } from '@/app/ui/products/product';
import Pagination from '../ui/products/pagination';
import { fetchProductsTotalPages } from '@/app/lib/data';
import Search from '../ui/products/search';

export default async function Page({
  searchParams
}: {
  searchParams?: {
    search?: string
    page?: string
  }
}) {
  const itemPerPage = 12;
  const search = searchParams?.search || '';
  const currentPage = Number(searchParams?.page) || 1;

  const totalPages = await fetchProductsTotalPages(itemPerPage, search);

  return (
    <>
      <h3 className="text-black text-3xl font-bold mt-3 leading-relaxed text-center mb-12 md:text-left md:mt-0 md:mb-12">
        Products
      </h3>

      <Search 
        placeholder="Search a product"
      />

      <AllProducts currentPage={currentPage} itemPerPage={itemPerPage} search={search} />

      <Pagination
        totalPages={totalPages}
      />
    </>
  );
}
