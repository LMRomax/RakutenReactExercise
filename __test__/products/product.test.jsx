import '@testing-library/jest-dom'
import { render, screen, cleanup } from '@testing-library/react'
import { Product } from '@/app/ui/products/product';

afterEach(() => {
    cleanup();
})

it('Should render product component in AllProducts', () => {
    const product = {
        id: 1,
        brand: 'Apple',
        title: 'Test Apple title',
        numprice: 20.99,
        usedprice: 12.50,
        href: '/url/to/test/apple',
        image: 'https://fr.shopping.rakuten.com/photo/4075345048_ML_NOPAD.jpg'
    };

    render(<Product
        id={product.id}
        brand={product.brand}
        title={product.title}
        numprice={product.numprice}
        usedprice={product.usedprice}
        href={product.href}
        image={product.image}
    />);
});

it('Should not render product component in AllProducts', () => {
    // if one data is null or undefined, product do not render
    const product = {
        id: 1,
        brand: null,
        title: 'Test Apple title',
        numprice: 20.99,
        usedprice: 12.50,
        href: '/url/to/test/apple',
        image: 'https://fr.shopping.rakuten.com/photo/4075345048_ML_NOPAD.jpg'
    };

    render(<Product
        id={product.id}
        brand={product.brand}
        title={product.title}
        numprice={product.numprice}
        usedprice={product.usedprice}
        href={product.href}
        image={product.image}
    />);
});

it('renders product unchanged', () => {
    const product = {
        id: 1,
        brand: 'Apple',
        title: 'Test Apple title',
        numprice: 20.99,
        usedprice: 12.50,
        href: '/url/to/test/apple',
        image: 'https://fr.shopping.rakuten.com/photo/4075345048_ML_NOPAD.jpg'
    };

    const { container } = render(<Product
        id={product.id}
        brand={product.brand}
        title={product.title}
        numprice={product.numprice}
        usedprice={product.usedprice}
        href={product.href}
        image={product.image}
    />)
    expect(container).toMatchSnapshot();
})
