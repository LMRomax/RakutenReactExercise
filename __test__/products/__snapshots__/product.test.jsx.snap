// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders product unchanged 1`] = `
<div>
  <a
    class="rounded-lg p-[16px] flex flex-row gap-[16px] items-center justify-center border border-slate-300 w-full"
    data-testid="product"
    href="/url/to/test/apple"
  >
    <div
      class="flex items-center justify-center w-[120px] h-[120px]"
    >
      <img
        alt="Image od product name"
        data-nimg="1"
        decoding="async"
        height="0"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Ffr.shopping.rakuten.com%2Fphoto%2F4075345048_ML_NOPAD.jpg&w=3840&q=75 3840w"
        style="color: transparent; width: 100%; height: auto;"
        width="0"
      />
    </div>
    <div>
      <div
        class="text-[#696969] text-[12px] font-semibold"
      >
        Apple
      </div>
      <div
        class="text-[#333] text-[14px] mb-[8px] font-semibold"
      >
        Test Apple title
      </div>
      <div
        class="text-[#BF0000] text-[14px] font-bold"
      >
        <span
          class="text-[20px]"
        >
          NaN €
        </span>
         Neuf
      </div>
      <div
        class="text-[#333] text-[14px]"
      >
        Occasion dés 
        <span
          class="text-[20px] font-bold"
        >
          NaN €
        </span>
      </div>
    </div>
  </a>
</div>
`;
